File: modules/app/agent-repl/daemon/internal/shimclient/control.go
package shimclient

import (
	"context"
	"errors"
	"fmt"
	"time"

	corev1 "agentrepl/proto/agentshim/core/v1"

	"claude-repld/internal/errclass"
	"claude-repld/internal/registry"

	"google.golang.org/protobuf/proto"
)

// The control-plane sentinels. Their VALUES live in internal/errclass, which
// is the package that classifies them; these are the historic names, kept so
// every existing errors.Is call site is unchanged and so the sentinel and its
// classification cannot be added independently of each other.

// ErrNotConnected is returned by a control send when there is no live shim
// connection. Control commands are never silently queued or dropped.
var ErrNotConnected = errclass.ErrShimNotConnected

// ErrNack is returned when the shim Nacks a control request. The reason is
// wrapped; there is NO silent retry (metaprompt no-fallbacks rule).
var ErrNack = errclass.ErrShimNack

// ErrAckTimeout is returned when no Ack/Nack arrives within the AckTimeout.
var ErrAckTimeout = errclass.ErrShimAckTimeout

// ErrDeliveredUnacked marks a control request that WAS WRITTEN to a live shim
// and whose connection then died before its receipt came back.
//
// IT IS A DIFFERENT FACT FROM EVERY OTHER CONTROL FAILURE, and the difference
// decides what the caller may conclude. ErrNotConnected means nothing left the
// daemon; a write error means the frame did not land; a Nack is the shim's own
// refusal. This one says the shim RECEIVED the request — the shim's dispatch is
// synchronous, so an interrupt that reached it has already run against the SDK —
// and only the answer was lost with the socket.
//
// So a caller must not treat it as "the command did not happen", and must not
// re-send to find out: re-sending an interrupt would stop a second, unrelated
// turn. What the outcome was is recovered from the reattach handshake and the
// durable store, never from this exchange.
var ErrDeliveredUnacked = errors.New("shimclient: control request was delivered but its receipt was lost with the connection")

// protoControl is a proto control message carrying a request_id.
type protoControl interface {
	proto.Message
	GetRequestId() string
}

// SubmitPrompt sends a prompt to the shim and blocks until the shim Acks or
// Nacks it (or ctx / the AckTimeout elapses). A Nack or timeout is a loud
// error — this layer never retries (the caller decides).
//
// requestID IS THE TURN'S IDENTITY, not merely a correlation token for the Ack.
// The shim adopts the accepted SubmitPrompt's request_id as the turn_id it
// stamps on TurnStarted, TurnEnded, Event.request_id and every consequence
// derived from them (core.proto §"Turn lifecycle authority"), so whatever id
// goes out here is the id the durable turn ledger, the keep-alive window row
// and SessionRewound.dropped_turn_ids will all name.
//
// A caller that already owns that identity — the keep-alive ping, whose claim,
// queue holds and window row are keyed by the id BEFORE the submit — passes it
// here, and the id it published is the id the turn comes back under. Minting a
// fresh one below would hand the same turn two names, and every match against
// the caller's name would then fail: exactly the defect that left keep-alive
// windows open forever and pings leaking into the live conversation.
//
// An empty requestID means the caller has no identity of its own for this turn
// and one is minted (newRequestID). That is not a fallback for a failed lookup:
// it is the honest statement that nothing daemon-side is keyed by this turn's
// name, so any unique name will do.
func (c *Client) SubmitPrompt(ctx context.Context, requestID, text, origin, permissionMode string, promptOrigin corev1.PromptOrigin) error {
	if promptOrigin == corev1.PromptOrigin_PROMPT_ORIGIN_UNSPECIFIED {
		return fmt.Errorf("shimclient: submit prompt requires a non-UNSPECIFIED prompt origin")
	}
	if _, ok := corev1.PromptOrigin_name[int32(promptOrigin)]; !ok {
		return fmt.Errorf("shimclient: submit prompt received unknown prompt origin %d", promptOrigin)
	}
	reqID := requestID
	if reqID == "" {
		minted, err := c.newRequestID("prompt")
		if err != nil {
			return err
		}
		reqID = minted
	}
	_, err := c.sendAwait(ctx, &corev1.SubmitPrompt{
		RequestId:      reqID,
		Text:           text,
		Origin:         origin,
		PermissionMode: permissionMode,
		PromptOrigin:   promptOrigin,
	}, reqID)
	return err
}

// Interrupt asks the shim to interrupt the current turn (the SDK's
// interrupt()) and awaits the Ack/Nack.
//
// It returns the shim's OWN verdict on what the stop did. That verdict is
// only obtainable here: from outside, a stop that failed and a turn that had
// already ended arrive as the same silence, which is how a no-op stop came to
// be painted as a failed one. A Nack or timeout is still an error, unchanged.
//
// originRequestID NAMES WHO ORDERED THE STOP, and it is carried purely so the
// log can say. A stop travels under a DAEMON-MINTED control id, which is the
// only id the wire exchange can be correlated by — and which appears nowhere in
// the vocabulary of whoever asked for it. A frontend interrupt was therefore
// unfindable: a search for the frontend's own request id turned up the command
// arriving and nothing else, so an interrupt that WAS delivered and acked read
// exactly like one the daemon had swallowed. Observed while diagnosing a stop
// that had in fact been answered INTERRUPTED within two milliseconds.
func (c *Client) Interrupt(ctx context.Context, originRequestID string) (corev1.InterruptOutcome, error) {
	reqID, err := c.newRequestID("interrupt")
	if err != nil {
		return corev1.InterruptOutcome_INTERRUPT_OUTCOME_UNSPECIFIED, err
	}
	ack, err := c.sendAwait(ctx, &corev1.Interrupt{RequestId: reqID}, originRequestID)
	if err != nil {
		return corev1.InterruptOutcome_INTERRUPT_OUTCOME_UNSPECIFIED, err
	}
	return ack.GetInterruptOutcome(), nil
}

// CancelDetachedAgents asks the shim to stop the session's DETACHED background
// agents — the subagent, shell and workflow tasks still working after the turn
// that launched them ended — and awaits the Ack/Nack.
//
// IT IS NOT Interrupt, AND THE DIFFERENCE IS THE WHOLE POINT. An Interrupt is
// aimed at the turn. The state a user reaches for this in is precisely the one
// where the turn is already over, so the interrupt is answered
// ALREADY_COMPLETE and the agents it was meant to stop are never touched.
//
// It returns the shim's OWN typed verdict, for the same reason Interrupt
// returns InterruptOutcome: only the shim can read its live-task set atomically
// against task end. A Nack or timeout is still an error, unchanged — the nil
// outcome that comes with one is never a "nothing was running" reading.
//
// originRequestID NAMES WHO ORDERED THE STOP and is carried purely so the log
// can say, exactly as Interrupt's is.
func (c *Client) CancelDetachedAgents(ctx context.Context, originRequestID string) (*corev1.DetachedCancelOutcome, error) {
	reqID, err := c.newRequestID("cancel-detached-agents")
	if err != nil {
		return nil, err
	}
	ack, err := c.sendAwait(ctx, &corev1.CancelDetachedAgents{RequestId: reqID}, originRequestID)
	if err != nil {
		return nil, err
	}
	outcome := ack.GetDetachedCancelOutcome()
	if outcome == nil {
		// AN ACK WITH NO OUTCOME IS A PROTOCOL VIOLATION, not an empty stop.
		// The contract sets the field on every arm — including both refusals —
		// so an ack without one is a shim that did something this daemon
		// cannot account for. Reading it as "nothing was running" would settle
		// no bubbles and report a successful cancel for work still in flight.
		return nil, fmt.Errorf("shim acked cancel-detached-agents request_id=%s with no detached_cancel_outcome; the contract sets it on every arm", reqID)
	}
	return outcome, nil
}

// QueryLiveTasks asks the live shim which background tasks it is running RIGHT
// NOW, and returns the shim's own list.
//
// A READ, NOT A CHANGE, and the daemon's only authority on the question. The
// task catalog is DERIVED from the events the daemon happened to see, so it can
// hold an entry open that nothing is running — a TaskEnded deduped as a replay,
// an end that landed while the daemon was down, a shim roll between the start
// and the end. This list is what tells that phantom apart from a subagent that
// is merely slow (sessioncontroller/phantomtask.go).
//
// AN ACK WITHOUT THE SET IS A FAILURE, NEVER AN EMPTY SET. The whole point of
// asking is that the answer is the shim's, so a peer that did not answer must
// not be read as one that answered "nothing" — a catalog entry would then be
// closed on silence, which is precisely what this path exists not to do. An ack
// carrying the message with no ids IS an answer, and the honest one.
func (c *Client) QueryLiveTasks(ctx context.Context) ([]string, error) {
	reqID, err := c.newRequestID("query-live-tasks")
	if err != nil {
		return nil, err
	}
	ack, nack, err := c.sendAwaitReceipt(ctx, &corev1.QueryLiveTasks{RequestId: reqID}, reqID)
	if err != nil {
		return nil, err
	}
	if nack != nil {
		c.logf("query-live-tasks REFUSED request_id=%s reason=%q", reqID, nack.GetReason())
		return nil, fmt.Errorf("%w: request_id=%s reason=%q", ErrNack, reqID, nack.GetReason())
	}
	set := ack.GetLiveTaskSet()
	if set == nil {
		return nil, fmt.Errorf("query live tasks: shim ack request_id=%s omitted live_task_set; an absent set is a peer that did not answer, not a session with no tasks", reqID)
	}
	ids := set.GetTaskIds()
	c.logf("query-live-tasks request_id=%s live_tasks=%d ids=%v", reqID, len(ids), ids)
	return ids, nil
}

// SetModel asks the live shim to change its SDK model and returns the model the
// shim confirmed. Callers persist only this value, never their requested one.
func (c *Client) SetModel(ctx context.Context, model string) (string, error) {
	requested := registry.NormalizeModel(model)
	if requested == "" {
		return "", fmt.Errorf("set model: model id is empty")
	}
	reqID, err := c.newRequestID("set-model")
	if err != nil {
		return "", err
	}
	ack, nack, err := c.sendAwaitReceipt(ctx, &corev1.SetModel{RequestId: reqID, Model: requested}, reqID)
	if err != nil {
		return "", err
	}
	if nack != nil {
		selected := registry.NormalizeModel(nack.GetSelectedModel())
		if selected == "" {
			return "", fmt.Errorf("set model: shim rejected request_id=%s without its current selected_model: %w", reqID, ErrNack)
		}
		c.logf("set-model REJECTED request_id=%s requested=%q selected=%q reason=%q", reqID, requested, selected, nack.GetReason())
		return selected, fmt.Errorf("%w: request_id=%s reason=%q", ErrNack, reqID, nack.GetReason())
	}
	selected := registry.NormalizeModel(ack.GetSelectedModel())
	if selected == "" {
		return "", fmt.Errorf("set model: shim ack request_id=%s omitted selected_model", reqID)
	}
	c.logf("set-model request_id=%s requested=%q selected=%q", reqID, requested, selected)
	return selected, nil
}

// QuerySelectedModel asks the live shim which model it is CURRENTLY running
// and returns the shim's own answer.
//
// A READ, NOT A CHANGE, and the only reason it exists is the bare
// argument-less `/model`: that form opens the CLI's own picker, which the
// daemon cannot stand in for, so without asking it would learn the resulting
// model only when some later submit re-announced a SystemInit. The picker
// would name the old model until then, and a hibernation before that would
// respawn the session on it.
//
// A SHIM THAT CANNOT ANSWER IS A HARD FAILURE, never an empty success. The
// caller's whole reason for asking is to commit a VERIFIED selection, so an
// empty answer is refused here rather than being written to a record as though
// somebody had confirmed it.
func (c *Client) QuerySelectedModel(ctx context.Context) (string, error) {
	reqID, err := c.newRequestID("query-selected-model")
	if err != nil {
		return "", err
	}
	ack, nack, err := c.sendAwaitReceipt(ctx, &corev1.QuerySelectedModel{RequestId: reqID}, reqID)
	if err != nil {
		return "", err
	}
	if nack != nil {
		c.logf("query-selected-model REFUSED request_id=%s reason=%q", reqID, nack.GetReason())
		return "", fmt.Errorf("%w: request_id=%s reason=%q", ErrNack, reqID, nack.GetReason())
	}
	selected := registry.NormalizeModel(ack.GetSelectedModel())
	if selected == "" {
		return "", fmt.Errorf("query selected model: shim ack request_id=%s omitted selected_model", reqID)
	}
	c.logf("query-selected-model request_id=%s selected=%q", reqID, selected)
	return selected, nil
}

// PermissionResponse sends the answer to an inbound PermissionRequest,
// correlated by the SAME request_id the shim used. It is fire-and-forget on
// the control plane (the shim uses it to unblock canUseTool); it is NOT
// Ack-awaited.
func (c *Client) PermissionResponse(resp *corev1.PermissionResponse) error {
	ac := c.currentConn()
	if ac == nil {
		return ErrNotConnected
	}
	if err := ac.writeMsg(resp); err != nil {
		return fmt.Errorf("sending PermissionResponse (request_id=%s): %w", resp.GetRequestId(), err)
	}
	c.logf("sent permission response request_id=%s decision=%s",
		resp.GetRequestId(), resp.GetDecision())
	return nil
}

// currentConn returns the live connection or nil.
func (c *Client) currentConn() *activeConn {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.active
}

// sendAwait registers a correlation waiter, sends msg, then blocks for the
// matching Ack/Nack. The waiter is always deregistered before returning.
//
// It returns the ACK itself, not just success: an Ack carries a verdict on
// some commands (an interrupt's outcome) and discarding it here would put the
// only process that can see that verdict in the position of having to guess
// at it later.
func (c *Client) sendAwait(ctx context.Context, msg protoControl, originRequestID string) (*corev1.Ack, error) {
	ack, nack, err := c.sendAwaitReceipt(ctx, msg, originRequestID)
	if err != nil {
		return nil, err
	}
	if nack != nil {
		return nil, fmt.Errorf("%w: request_id=%s reason=%q", ErrNack, nack.GetRequestId(), nack.GetReason())
	}
	return ack, nil
}

// sendAwaitReceipt is the shared correlated-control exchange.  SetModel needs
// a rejected receipt's shim-confirmed selected_model; legacy controls retain
// sendAwait's simpler Ack-or-error contract above.
//
// originRequestID is the id of whatever ORDERED this exchange, in its own
// vocabulary, and it is logged beside the daemon-minted control id on both the
// send and the ack. The two ids are the only bridge between a caller's record
// and the wire, and a caller with no id of its own passes "" — which the log
// then renders as an exchange nothing outside this package named.
func (c *Client) sendAwaitReceipt(ctx context.Context, msg protoControl, originRequestID string) (*corev1.Ack, *corev1.Nack, error) {
	reqID := msg.GetRequestId()
	ac := c.currentConn()
	if ac == nil {
		return nil, nil, ErrNotConnected
	}

	ch := make(chan ackResult, 1)
	ac.pendMu.Lock()
	ac.pending[reqID] = ch
	ac.pendMu.Unlock()
	defer func() {
		ac.pendMu.Lock()
		delete(ac.pending, reqID)
		ac.pendMu.Unlock()
	}()

	if err := ac.writeMsg(msg); err != nil {
		return nil, nil, fmt.Errorf("sending control request (request_id=%s): %w", reqID, err)
	}
	c.logf("sent control request request_id=%s origin_request_id=%q, awaiting ack (timeout=%s)", reqID, originRequestID, c.cfg.AckTimeout)

	timer := time.NewTimer(c.cfg.AckTimeout)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return nil, nil, fmt.Errorf("control request %s cancelled: %w", reqID, ctx.Err())
	case <-timer.C:
		return nil, nil, fmt.Errorf("%w: request_id=%s after %s", ErrAckTimeout, reqID, c.cfg.AckTimeout)
	case res := <-ch:
		if res.err != nil {
			// The write above SUCCEEDED and the connection then died, so the
			// shim has the request and the daemon has lost only its answer.
			// Named as such, because the caller's next move depends entirely on
			// which of those two it was (see ErrDeliveredUnacked).
			c.logf("control request request_id=%s origin_request_id=%q DELIVERED but its receipt was lost with the connection: %v — the shim ran it; the outcome is recovered from the reattach, never by re-sending",
				reqID, originRequestID, res.err)
			return nil, nil, fmt.Errorf("%w: request_id=%s: %w", ErrDeliveredUnacked, reqID, res.err)
		}
		if res.nack != nil {
			return nil, res.nack, nil
		}
		c.logf("control request request_id=%s origin_request_id=%q acked outcome=%s", reqID, originRequestID, res.ack.GetInterruptOutcome())
		return res.ack, nil, nil
	}
}

// resolveAck delivers an Ack to the correlated waiter. An Ack with no waiter is
// loud-logged (a stray or duplicate ack) but not fatal.
func (c *Client) resolveAck(ac *activeConn, ack *corev1.Ack) {
	if !ac.deliver(ack.GetRequestId(), ackResult{ack: ack}) {
		c.logf("received Ack for unknown request_id=%s (stray or late)", ack.GetRequestId())
	}
}

// resolveNack delivers a Nack (loud) to the correlated waiter.
func (c *Client) resolveNack(ac *activeConn, nack *corev1.Nack) {
	c.logf("received Nack request_id=%s reason=%q", nack.GetRequestId(), nack.GetReason())
	if !ac.deliver(nack.GetRequestId(), ackResult{nack: nack}) {
		c.logf("received Nack for unknown request_id=%s (stray or late)", nack.GetRequestId())
	}
}

// deliver hands res to the waiter for reqID, returning whether one existed.
func (ac *activeConn) deliver(reqID string, res ackResult) bool {
	ac.pendMu.Lock()
	ch, ok := ac.pending[reqID]
	if ok {
		delete(ac.pending, reqID)
	}
	ac.pendMu.Unlock()
	if !ok {
		return false
	}
	ch <- res
	return true
}

File: modules/app/agent-repl/agent-shim/claude/shim/src/uds/control.ts
/**
 * Control-plane glue between the daemon connection and the SDK session.
 *
 * Two responsibilities:
 * 1. Map inbound `SubmitPrompt` / `Interrupt` onto an injected SDK-session
 *    interface ({@link SdkControlTarget}) and return the synchronous
 *    `Ack`/`Nack` receipt the server writes back.
 * 2. Drive the `canUseTool` PermissionRequest round-trip: {@link requestPermission}
 *    emits a `PermissionRequest` to the daemon and BLOCKS (returns a Promise)
 *    on a pending-request map keyed by `request_id` until the matching
 *    `PermissionResponse` arrives via {@link handlePermissionResponse}.
 *    The shim is the DURABLE OWNER of that open question: while an answer is
 *    outstanding the request is re-sent on every reattach
 *    ({@link resendPending}), so neither an unattached daemon nor a daemon that
 *    died holding the ask can strand the turn.
 *
 * This deliberately does NOT import src/session.ts. {@link SdkControlTarget}
 * is the minimal seam the stitch phase binds to the real ShimSession.
 */
import { randomUUID } from "node:crypto";
import { create } from "@bufbuild/protobuf";
import type { JsonObject } from "@bufbuild/protobuf";
import { bindLog } from "./log.js";
import { normalizeModel } from "../model.js";
import {
  Ack,
  AckSchema,
  Interrupt,
  InterruptOutcome,
  Nack,
  NackSchema,
  PermissionDecision,
  PermissionRequest,
  PermissionRequestSchema,
  PermissionResponse,
  PromptOrigin,
  PromptOriginSchema,
  QuerySelectedModel,
  SetModel,
  SubmitPrompt,
} from "./proto.js";

/**
 * The minimal SDK-session surface control dispatch drives. The stitch phase
 * binds this to the real ShimSession (streaming-input push + interrupt).
 */
export interface SdkControlTarget {
  /** Push one user prompt into the SDK streaming-input turn. */
  submitPrompt(input: {
    requestId: string;
    text: string;
    origin: string;
    promptOrigin: PromptOrigin;
    permissionMode?: string;
  }): Promise<void>;
  /**
   * Interrupt the current turn via the SDK `interrupt()`.
   *
   * Returns the THREE-VALUED outcome, decided SYNCHRONOUSLY. The
   * implementation owns turn lifecycle, so only it can answer "was a turn
   * live when this landed?" without racing — see the emitter in
   * uds-session.ts for why that read must precede every await.
   */
  interrupt(input: { requestId: string }): InterruptOutcome;
  /** Set a model and return the SDK-confirmed selection. */
  setModel(input: { requestId: string; model: string }): Promise<string>;
  /**
   * The model this session is CURRENTLY running.
   *
   * A READ, not a change, and synchronous: the implementation already holds
   * the answer — it tracks every SDK `system:init` and every confirmed
   * `setModel` — so there is nothing to await and nothing to race. Empty means
   * the shim has never observed a real model, which is a refusal rather than
   * an answer (see handleQuerySelectedModel).
   */
  selectedModel(): string;
}

/**
 * A rejected model mutation can still report the SDK's current selection.
 * That value is authoritative state, not an optimistic echo of the request.
 */
export class ModelSelectionError extends Error {
  constructor(message: string, readonly selectedModel: string) {
    super(message);
    this.name = "ModelSelectionError";
  }
}

/** The shim's own resolution of a canUseTool request. */
export type ToolPermissionResult =
  | { behavior: "allow"; updatedInput: JsonObject }
  | { behavior: "deny"; message: string };

/**
 * Sends a PermissionRequest to the daemon (bound to SessionServer).
 *
 * Returns whether the frame reached a live daemon connection. `false` is NOT a
 * failure to be swallowed: the request stays pending here and is re-sent by
 * {@link ControlDispatch.resendPending} on the next reattach.
 */
export type PermissionRequestSender = (req: PermissionRequest) => boolean;

export interface ControlDispatchOptions {
  /** Request-id minter; defaults to randomUUID. Injectable for tests. */
  newRequestId?: () => string;
  /**
   * Whether an SDK turn is currently in flight. A pending permission belongs to
   * the turn whose tool call raised it, so a re-send with no live turn would be
   * re-asking a dead turn's question — {@link ControlDispatch.resendPending}
   * cancels those waits instead of putting them back in front of the user.
   *
   * Omitted, liveness is unknown and re-sends are unconditional (the old
   * behavior, and the right default for a dispatch with no turn owner bound).
   */
  isTurnLive?: () => boolean;
}

interface PendingPermission {
  resolve: (result: ToolPermissionResult) => void;
  input: JsonObject;
  /**
   * The EXACT frame first sent for this request. A re-send replays this value
   * rather than rebuilding one, so the identity the daemon rendezvouses on
   * (request_id) and everything it renders (tool name, input) are bit-identical
   * across every attempt.
   */
  request: PermissionRequest;
}

const COMPONENT = "shim-control";
const LOGGER = bindLog({ component: COMPONENT, operation: "shim.control.dispatch" });

export class ControlDispatch {
  private readonly newRequestId: () => string;
  private readonly isTurnLive: (() => boolean) | undefined;
  /**
   * Insertion-ordered by construction (Map iteration order), which is what
   * makes a multi-request re-send arrive in the order the user was originally
   * asked.
   */
  private readonly pending = new Map<string, PendingPermission>();

  constructor(
    private readonly target: SdkControlTarget,
    private readonly sendPermissionRequest: PermissionRequestSender,
    opts: ControlDispatchOptions = {},
  ) {
    this.newRequestId = opts.newRequestId ?? randomUUID;
    this.isTurnLive = opts.isTurnLive;
  }

  /** Handle a SubmitPrompt; push into the SDK turn and Ack (Nack on rejection). */
  async handleSubmitPrompt(msg: SubmitPrompt): Promise<Ack | Nack> {
    if (msg.requestId.trim() === "") {
      const reason = "SubmitPrompt requires a non-empty request_id";
      LOGGER.log({ level: "error" }, reason);
      return create(NackSchema, { requestId: msg.requestId, reason });
    }
    LOGGER.log({ request_id: msg.requestId, origin: msg.origin, prompt_origin: msg.promptOrigin, text_length: msg.text.length, permission_mode: msg.permissionMode || undefined }, "dispatching SubmitPrompt to SDK session");
    if (msg.promptOrigin === PromptOrigin.UNSPECIFIED || !PromptOriginSchema.values.some((value) => value.number === msg.promptOrigin)) {
      const reason = `SubmitPrompt ${JSON.stringify(msg.requestId)} has invalid prompt_origin ${msg.promptOrigin}`;
      LOGGER.log({ level: "error", request_id: msg.requestId, prompt_origin: msg.promptOrigin }, reason);
      return create(NackSchema, { requestId: msg.requestId, reason });
    }
    try {
      await this.target.submitPrompt({
        requestId: msg.requestId,
        text: msg.text,
        origin: msg.origin,
        promptOrigin: msg.promptOrigin,
        ...(msg.permissionMode !== "" ? { permissionMode: msg.permissionMode } : {}),
      });
      LOGGER.log({ request_id: msg.requestId }, "SubmitPrompt accepted by SDK session");
      return create(AckSchema, { requestId: msg.requestId });
    } catch (err) {
      LOGGER.log({ level: "error", request_id: msg.requestId, cause: err }, `submit-prompt failed: ${errMsg(err)}`);
      return create(NackSchema, { requestId: msg.requestId, reason: errMsg(err) });
    }
  }

  /**
   * Handle an Interrupt; forward to the SDK turn and Ack with the outcome
   * the target decided (Nack on throw).
   *
   * THE AMBIGUITY THIS REMOVES. A consumer watching for an `aborted` result
   * sees the turn end and the interrupt arrive as two unordered events, and
   * so cannot tell a stop that FAILED from a turn that had ALREADY ENDED —
   * it painted the second as the first. The shim owns turn lifecycle and its
   * event loop is single-threaded, so the answer is unambiguous HERE and
   * nowhere downstream. Carrying it on the Ack is what stops every consumer
   * from re-deriving it wrongly.
   *
   * A synchronous throw stays a NACK rather than becoming an Ack carrying
   * FAILED: a Nack is the stronger, already-established failure signal on
   * this wire, and downgrading it to a successful receipt with a sad field
   * would weaken error coverage the daemon already acts on. FAILED is for
   * the case the target can prove undeliverable while still answering.
   */
  handleInterrupt(msg: Interrupt): Ack | Nack {
    LOGGER.log({ request_id: msg.requestId }, "dispatching Interrupt to SDK session");
    try {
      const outcome = this.target.interrupt({ requestId: msg.requestId });
      LOGGER.log({ request_id: msg.requestId, interrupt_outcome: outcome }, "Interrupt resolved by SDK session");
      return create(AckSchema, { requestId: msg.requestId, interruptOutcome: outcome });
    } catch (err) {
      LOGGER.log({ level: "error", request_id: msg.requestId, cause: err }, `interrupt failed: ${errMsg(err)}`);
      return create(NackSchema, { requestId: msg.requestId, reason: errMsg(err) });
    }
  }

  /** Apply a deliberate model request only after the SDK settles it. */
  async handleSetModel(msg: SetModel): Promise<Ack | Nack> {
    const model = normalizeModel(msg.model);
    if (model === "") {
      const reason = "set-model requires a non-empty model id; <synthetic> is not selectable";
      LOGGER.log({ level: "error", request_id: msg.requestId, requested_model: msg.model }, reason);
      return create(NackSchema, { requestId: msg.requestId, reason });
    }
    try {
      const selectedModel = normalizeModel(await this.target.setModel({ requestId: msg.requestId, model }));
      if (selectedModel === "") {
        throw new Error("SDK acknowledged SetModel without a real selected model");
      }
      return create(AckSchema, { requestId: msg.requestId, selectedModel });
    } catch (err) {
      const reason = errMsg(err);
      const selectedModel = err instanceof ModelSelectionError ? normalizeModel(err.selectedModel) : "";
      LOGGER.log({ level: "error", request_id: msg.requestId, model, selected_model: selectedModel }, `set-model failed: ${reason}`);
      return create(NackSchema, { requestId: msg.requestId, reason, selectedModel });
    }
  }

  /**
   * Answer which model this session is currently running.
   *
   * WHY THE DAEMON ASKS. The bare argument-less `/model` opens the CLI's own
   * picker, which no layer above can intercept: without this read the daemon
   * learns the resulting model only when some later submit happens to
   * re-announce a `system:init`, so the topbar picker names the OLD model
   * until then and a hibernation before that respawns the session on it.
   *
   * AN UNKNOWN MODEL IS A NACK, NEVER AN EMPTY ACK. An Ack carrying no
   * selection would be indistinguishable from a confirmed absence, and the
   * daemon would have to guess which it was; the whole point of the read is
   * that the value it commits was verified. A refusal is the honest answer and
   * the daemon surfaces it as a failure.
   */
  handleQuerySelectedModel(msg: QuerySelectedModel): Ack | Nack {
    try {
      const selectedModel = normalizeModel(this.target.selectedModel());
      if (selectedModel === "") {
        const reason = "query-selected-model: this session has not observed a real model yet";
        LOGGER.log({ level: "error", request_id: msg.requestId }, reason);
        return create(NackSchema, { requestId: msg.requestId, reason });
      }
      LOGGER.log({ request_id: msg.requestId, selected_model: selectedModel }, "reported the live selected model");
      return create(AckSchema, { requestId: msg.requestId, selectedModel });
    } catch (err) {
      const reason = errMsg(err);
      LOGGER.log({ level: "error", request_id: msg.requestId, cause: err }, `query-selected-model failed: ${reason}`);
      return create(NackSchema, { requestId: msg.requestId, reason });
    }
  }

  /**
   * The SDK canUseTool callback. Emits a PermissionRequest to the daemon and
   * resolves only when the matching PermissionResponse arrives (or the
   * request is cancelled). One entry per request_id in the pending map.
   */
  requestPermission(toolName: string, input: JsonObject): Promise<ToolPermissionResult> {
    const requestId = this.newRequestId();
    const request = create(PermissionRequestSchema, { requestId, toolName, input });
    return new Promise<ToolPermissionResult>((resolve, reject) => {
      this.pending.set(requestId, { resolve, input, request });
      LOGGER.log({ request_id: requestId, tool_name: toolName, pending_count: this.pending.size, input_keys: Object.keys(input) }, "permission request opened for daemon decision");
      try {
        if (!this.sendPermissionRequest(request)) {
          // UNDELIVERED, NOT LOST. There is no daemon attached to answer, so
          // the ask is retained here and re-sent by resendPending the moment
          // one reattaches. Rejecting instead would fail the tool call for a
          // transport gap the user never chose.
          LOGGER.log({ level: "error", request_id: requestId, tool_name: toolName, pending_count: this.pending.size },
            "permission request could not be delivered (no daemon attached); it stays pending and is re-sent on reattach");
        }
      } catch (err) {
        this.pending.delete(requestId);
        LOGGER.log({ level: "error", request_id: requestId, tool_name: toolName, input_keys: Object.keys(input), pending_count: this.pending.size, cause: err }, `permission request send failed: ${errMsg(err)}`);
        reject(err);
      }
    });
  }

  /**
   * Re-send every still-unanswered permission request to the daemon.
   *
   * Called once per completed reattach (after ShimReady closes the bring-up
   * gate). It closes BOTH halves of the rendezvous gap: a request raised while
   * no daemon was attached was never delivered, and a request the previous
   * daemon received died with that daemon's in-memory rendezvous. Neither the
   * daemon nor the frontend can recover the question — only the side still
   * holding the blocked `canUseTool` promise can, and that is this map.
   *
   * WHY THE SAME FRAME AND THE SAME request_id. The re-send is not a new ask;
   * it is the same one, restated. Identity is what lets the daemon re-arm
   * idempotently: an id it has no waiter for re-establishes the wait, an id it
   * already answered replays the recorded answer instead of asking a human the
   * same question twice.
   *
   * NO INTERLEAVING WITH AN ANSWER. This is synchronous over the map on a
   * single-threaded event loop, so an answer that arrives just before it runs
   * has already removed its entry, and one that arrives just after resolves a
   * request whose re-send is merely redundant. Either way exactly one answer
   * resolves the wait.
   *
   * A DEAD TURN IS CANCELLED, NOT RE-ASKED. If no SDK turn is in flight the
   * tool call these requests block no longer exists, so re-asking would put a
   * question in front of the user that nothing is waiting on. Those waits are
   * force-denied through the same path an interrupt uses.
   */
  resendPending(reason: string): void {
    if (this.pending.size === 0) return;
    if (this.isTurnLive !== undefined && !this.isTurnLive()) {
      this.cancelAll(`${reason}: no turn is in flight to receive the answer`);
      return;
    }
    LOGGER.log({
      pending_count: this.pending.size,
      reason,
      request_ids: [...this.pending.keys()],
    }, "re-sending every unanswered permission request to the reattached daemon");
    for (const [requestId, pending] of this.pending) {
      try {
        if (!this.sendPermissionRequest(pending.request)) {
          LOGGER.log({ level: "error", request_id: requestId, tool_name: pending.request.toolName, reason },
            "permission request re-send found no daemon attached; it stays pending for the next reattach");
        }
      } catch (err) {
        // A throwing send is a transport fault, not an answer. The request
        // stays pending so the next reattach re-sends it; failing the tool
        // call here would decide on the user's behalf.
        LOGGER.log({ level: "error", request_id: requestId, tool_name: pending.request.toolName, reason, cause: err },
          `permission request re-send failed: ${errMsg(err)}`);
      }
    }
  }

  /** Resolve the blocked canUseTool round-trip a PermissionResponse targets. */
  handlePermissionResponse(msg: PermissionResponse): void {
    const pending = this.pending.get(msg.requestId);
    if (!pending) {
      LOGGER.log({ level: "warn", request_id: msg.requestId }, `permission-response for unknown request_id (ignored)`);
      return;
    }
    this.pending.delete(msg.requestId);
    if (msg.decision === PermissionDecision.ALLOW) {
      // Allow-with-edits: updated_input, when present, replaces the input.
      const updatedInput = msg.updatedInput ?? pending.input;
      LOGGER.log({ request_id: msg.requestId, decision: "allow", edited: msg.updatedInput !== undefined, pending_count: this.pending.size }, "permission request resolved");
      pending.resolve({ behavior: "allow", updatedInput });
    } else {
      LOGGER.log({ request_id: msg.requestId, decision: "deny", pending_count: this.pending.size }, "permission request resolved");
      pending.resolve({
        behavior: "deny",
        message: msg.denyMessage !== "" ? msg.denyMessage : "permission denied",
      });
    }
  }

  /**
   * Cancel every outstanding permission wait (interrupt / disconnect /
   * shutdown), resolving each as a deny so no SDK callback hangs forever.
   */
  cancelAll(reason: string): void {
    if (this.pending.size > 0) {
      // Each cancelled wait is a tool DENIED without the user ever being
      // asked, so this is a decision made on their behalf rather than a
      // lifecycle note. The request ids go with it: the bounded set is what
      // lets an operator tell which asks were answered for them.
      LOGGER.log({
        level: "warn",
        pending_count: this.pending.size,
        reason,
        request_ids: [...this.pending.keys()],
      }, "cancelling outstanding permission requests; each is force-denied without the user seeing the prompt");
    }
    for (const [id, pending] of this.pending) {
      this.pending.delete(id);
      pending.resolve({ behavior: "deny", message: reason });
    }
  }

  /** Count of in-flight permission round-trips (for tests / diagnostics). */
  pendingCount(): number {
    return this.pending.size;
  }
}

function errMsg(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}
